### PR TITLE
Add repository-url module and move URLRepository

### PR DESCRIPTION
**Pull request by Tim-Brooks**
This is related to #22116. `URLRepository` requires `SocketPermission` 
`connect`. This commit introduces a new module called "repository-url" 
where `URLRepository` will reside. With the new module, permissions can 
be removed from core.

### CODE REVIEW

**Comment by rjernst** [comment]
Is there any reason FSRepository would need to be in the same module as URL repository, or could it be separate when it is eventually moved? In which case, this can be called `repository-url`?

File: core/src/test/java/org/elasticsearch/bwcompat/RestoreBackwardsCompatIT.java [hunk 1]
```diff
@@ -98,7 +91,8 @@ public void testRestoreOldSnapshots() throws Exception {
         for (Version v : VersionUtils.allReleasedVersions()) {
             if (VersionUtils.isSnapshot(v)) continue;  // snapshots are unreleased, so there is no backcompat yet
             if (v.isRelease() == false) continue; // no guarantees for prereleases
-            if (v.before(Version.CURRENT.minimumIndexCompatibilityVersion())) continue; // we only support versions N and N-1
+            if (v.before(Version.CURRENT.minimumIndexCompatibilityVersion()))
```
Comment (rjernst) [inline core/src/test/java/org/elasticsearch/bwcompat/RestoreBackwardsCompatIT.java]: Why this change? If you are going to move the continue to the next line (breaking with the convention of the other loop shortcut statements here...), please open a block.

File: modules/repositories/build.gradle [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+esplugin {
+    description 'Module for generic repositories'
+    classname 'org.elasticsearch.plugin.repositories.RepositoriesPlugin'
+}
+
+dependencies {
```
Comment (rjernst) [inline modules/repositories/build.gradle]: If there are no dependencies, you don't need a dependencies section.

File: modules/repositories/build.gradle [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+}
+
+compileJava.options.compilerArgs << "-Xlint:-unchecked,-rawtypes"
+compileTestJava.options.compilerArgs << "-Xlint:-unchecked,-rawtypes"
+
+thirdPartyAudit.excludes = [
```
Comment (rjernst) [inline modules/repositories/build.gradle]: Not necessary.

File: modules/repositories/src/main/java/org/elasticsearch/plugin/repositories/RepositoriesPlugin.java [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+        return Arrays.asList(
+            URLRepository.ALLOWED_URLS_SETTING,
+            URLRepository.REPOSITORIES_URL_SETTING,
+            URLRepository.SUPPORTED_PROTOCOLS_SETTING
+        );
+    }
+
+    @Override
+    public Map<String, Repository.Factory> getRepositories(Environment env, NamedXContentRegistry namedXContentRegistry) {
+        return Collections.singletonMap(URLRepository.TYPE, (metadata) -> new URLRepository(metadata, env, namedXContentRegistry));
```
Comment (rjernst) [inline modules/repositories/src/main/java/org/elasticsearch/plugin/repositories/RepositoriesPlugin.java]: nit: the parens around `metadata` are not necessary.

**Review by rjernst — COMMENTED** [review]
This looks good. I left a couple minor comments. The main thing I want to clear up is the name.

File: core/src/test/java/org/elasticsearch/bwcompat/RestoreBackwardsCompatIT.java (hunk 1 above)
Comment (tbrooks8) [inline core/src/test/java/org/elasticsearch/bwcompat/RestoreBackwardsCompatIT.java]: I'm not sure. I definitely did not intend that change. I'll revert that line.

**Comment by tbrooks8** [comment]
I'll update the PR with the requested changes.

As for the name - when @s1monw and I originally talked about it, I was going to move both to a module named "repositories". But then [this](https://github.com/elastic/elasticsearch/issues/22116#issuecomment-273675490) kind of blocked that from being a straightforward change.

So we agreed that while working on the `SocketPermissions` work I would only move the `URLRepository`. I kept the naming with the assumption that at some point we would move the `FSRepository` there also.

But there is no technical reason that I know of why they would need to be in the same module. If we want to specialize this module I can make those changes.

**Comment by rjernst** [comment]
> But there is no technical reason that I know of why they would need to be in the same module.

Since URLRepository will require network permission, but FSRepository will not, I think that would be a good split.

**Comment by jasontedor** [comment]
> Since URLRepository will require network permission, but FSRepository will not, I think that would be a good split.

+1

**Comment by tbrooks8** [comment]
@s1monw are you alright with the name change?

**Comment by s1monw** [comment]
> @s1monw are you alright with the name change?

sure

**Comment by s1monw** [comment]
code LGTM - let do the name change and then get this in? I also wonder if we wanna add the permission already in this PR of if you wanna add it all together? either way is find

**Review by s1monw — APPROVED** [review]

**Comment by tbrooks8** [comment]
I renamed it to `url-repository`.

I also worked on fixing some integration tests. I moved a snapshot restore test that used the `URLRepository` to the new module. That test required a repository path and it looks like you can only call `randomRepoPath()` on `ESIntegTestCase` if you're an internal cluster?

In order to force that I a method I had to do [this](https://github.com/elastic/elasticsearch/pull/22752/files#diff-85c8f575c2f556c0f8fdfae7ac1e5da5R47). I'm not sure if this was the correct approach? This was the first time I was dealing with the integration test harness, so let me know if there is something I should be doing instead.

**Comment by jasontedor** [comment]
> I renamed it to `url-repository`.

I'm really sorry to be a pain here, but I think that it should be `repository-url` (we have plugins already named `repository-*`).

**Comment by tbrooks8** [comment]
@jasontedor I mistyped the comment. `repository-url` is actually what I named it.

**Comment by jasontedor** [comment]
> I mistyped the comment. repository-url is actually what I named it.

Thank you. :smile:

**Comment by rjernst** [comment]
@tbrooks8 I don't see where your `ignoreExternalCluster()` is actually called? But ESIntegTestCase subclasses to not use an external cluster (anymore, they did long ago), so I'm confused on what instigated this added boolean?

**Comment by tbrooks8** [comment]
@rjernst Right [here](https://github.com/elastic/elasticsearch/blob/master/test/framework/src/main/java/org/elasticsearch/test/ESIntegTestCase.java#L1762). 

So when I run gradle :modules:repository-url:integTest without that , this [exception](https://github.com/elastic/elasticsearch/blob/master/test/framework/src/main/java/org/elasticsearch/test/ESIntegTestCase.java#L1937) is being thrown.

**Comment by rjernst** [comment]
I think you need to rename that test. That exception is there for a reason. It means the test is running with an external cluster (which integTest does everywhere _except_ in core, where the test was pulled from). In core, tests ending in IT extend from ESIntegTestCase. For all other modules, tests ending in IT mean they are running against an external cluster (a true integ test). If you rename the suffix of that test to `Tests`, it should work like a normal core integ test.

**Comment by tbrooks8** [comment]
Thanks. That looks like it was my issue.

**Comment by tbrooks8** [comment]
@jasontedor / @rjernst It looks like the build is passing at this point (now that I have worked out integration test issues). Can you update your reviews and approve it if everything looks alright?

File: modules/repository-url/build.gradle [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+esplugin {
+    description 'Module for generic repositories'
```
Comment (rjernst) [inline modules/repository-url/build.gradle]: Should be URL repository?

**Review by rjernst — APPROVED** [review]
LGTM

File: modules/repository-url/build.gradle (hunk 5 above)
Comment (tbrooks8) [inline modules/repository-url/build.gradle]: Yep - just fixed

File: modules/repository-url/build.gradle [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+ * under the License.
+ */
+
+esplugin {
+    description 'Module for URL repository'
+    classname 'org.elasticsearch.plugin.repository.url.URLRepositoryPlugin'
+}
+
+compileJava.options.compilerArgs << "-Xlint:-unchecked,-rawtypes"
+compileTestJava.options.compilerArgs << "-Xlint:-unchecked,-rawtypes"
```
Comment (jasontedor) [inline modules/repository-url/build.gradle]: Can you explain why these are needed?

**Review by jasontedor — COMMENTED** [review]
Looks great, I left a question though.

File: modules/repository-url/build.gradle (hunk 6 above)
Comment (tbrooks8) [inline modules/repository-url/build.gradle]: I just copied an existing module and made appropriate modifications when I created this. I assume those are artifacts from that and I do not know any reason those are need. I'll remove them.

Comment (jasontedor) [inline modules/repository-url/build.gradle]: Cool, let's indeed see if we can drop them.

**Review by jasontedor — APPROVED** [review]
LGTM. If the CI build passes, I'm good with it.

**Comment by tbrooks8** [comment]
The CI build passed. So I'll merge this.
